Wishlist Service

legendofa opened the pull request.

nk-coding commented on compose.yaml:
```diff
@@ -0,0 +1,18 @@
+services:
+  app:
```
this one currently does not work yet

nk-coding commented on src/mutation_root.rs:
```diff
@@ -0,0 +1,78 @@
+use crate::wishlist::Wishlist;
+pub struct MutationRoot;
+
+fn uuid_parse(id: &String) -> Result<String, Error> {
```
missing documentation

legendofa commented:
Shall i also wrap mutations with only one argument, as delete, in a `DeleteWishlistInput` object?

nk-coding commented:
> Shall i also wrap mutations with only one argument, as delete, in a `DeleteWishlistInput` object?

Delete is a hard one. I do not really have a fixed opinion about this.  For create and update with one input, I would still say yes. Regarding delete, a benefit would be consistency, however I do also see why one would argue this is completely redundant.
One argument could be that in some cases, even delete could need more arguments, .e.g. a flag if some other entities should be deleted with the entity.
In any case, we should come to a decision we follow in all services.

legendofa commented:
Why should created/update return the object, isn't that overhead? Also did not find out yet how to implement pagination

legendofa commented:
> > Shall i also wrap mutations with only one argument, as delete, in a `DeleteWishlistInput` object?
> 
> Delete is a hard one. I do not really have a fixed opinion about this. For create and update with one input, I would still say yes. Regarding delete, a benefit would be consistency, however I do also see why one would argue this is completely redundant. One argument could be that in some cases, even delete could need more arguments, .e.g. a flag if some other entities should be deleted with the entity. In any case, we should come to a decision we follow in all services.

:+1:  I just left it concise for now, we can decide this in the next meeting then :)

nk-coding commented:
> Why should created/update return the object, isn't that overhead? Also did not find out yet how to implement pagination

First, for create, if you don't return it, how should the frontend know the id of the created object?
Second, often, when you create or update sth, you want to display the created/updated thing. Returning it allows you to do so without having to make another query.
Also, regarding efficiency, there should not be too much overhead. Returning it allows the user to fetch information from the created/update entity, however, GraphQL does not require you to do so (except at least one scalar field, but that is minimal overhead imho).

legendofa commented:
> First, for create, if you don't return it, how should the frontend know the id of the created object? Second, often, when you create or update sth, you want to display the created/updated thing. Returning it allows you to do so without having to make another query. Also, regarding efficiency, there should not be too much overhead. Returning it allows the user to fetch information from the created/update entity, however, GraphQL does not require you to do so (except at least one scalar field, but that is minimal overhead imho).

Sounds good, i have adapted it to return the UUID on creation and modification now :)

nk-coding commented:
> > First, for create, if you don't return it, how should the frontend know the id of the created object? Second, often, when you create or update sth, you want to display the created/updated thing. Returning it allows you to do so without having to make another query. Also, regarding efficiency, there should not be too much overhead. Returning it allows the user to fetch information from the created/update entity, however, GraphQL does not require you to do so (except at least one scalar field, but that is minimal overhead imho).
> 
> Sounds good, i have adapted it to return the UUID on creation and modification now :)

Again, it would be better to return the whole object to allow the user to decide what to get. In GraphQL, you basically never return just an id.
We should probably discuss if we want to support the Payload pattern (wrap the output in a Payload object).
See https://blog.logrocket.com/anti-patterns-graphql-schema-design/ (insufficient mutation response)

nk-coding reviewed (changes requested):
- WishlistOrder should have fields `direction` and `field` to be consistent with other servics
- Missing documentation for WishlistOrder, WishlistOrderField, OrderDirection
- QueryRoot should be named Query (federation support)
- MutationRoot should be named Mutation (federation support)
- if possible, use ID scalar instead of custom UUID scalar (maybe just rename)
- wishlists query should return a WishlistConnection which also allows to get totalCount, hasNextPage
- field inputs are often not documented, instead the documentation is part of the field itself (however there is also documentation for ctx which does not exist in GraphQL)

legendofa commented:
Would it be possible to use UUIDs directly, as IDs lead to more error handling stuff? We only use UUIDs so no general IDs should be required

legendofa commented:
>     * field inputs are often not documented, instead the documentation is part of the field itself (however there is also documentation for ctx which does not exist in GraphQL)

Removed the `ctx`, do we really need duplicate docs there? The inputs are already the documented input object.

nk-coding commented:
> >     * field inputs are often not documented, instead the documentation is part of the field itself (however there is also documentation for ctx which does not exist in GraphQL)
> 
> Removed the `ctx`, do we really need duplicate docs there? The inputs are already the documented input object.
> 

Well if I could choose, I would add it only to the inputs themselves.

legendofa commented:
I have added Connection support in this PR. Changed the field docs to the annotation style, but i am not sure if it turns out the way we want it to be. Next up will be the federation support, but will create a new PR for that.

nk-coding reviewed (changes requested):
- In GraphQL, maybe rename BaseConnection to WishlistConnection as it is a connection for Wishlists (also we will have clashes otherwise in the future).
- BaseConnection and Wishlist are both missing GraphQL descriptions for all fields

legendofa commented:
> * In GraphQL, maybe rename BaseConnection to WishlistConnection as it is a connection for Wishlists (also we will have clashes otherwise in the future).
> 
>     * BaseConnection and Wishlist are both missing GraphQL descriptions for all fields

I am aware of this and did not find out how to rename it, there is no such thing as inheritance in Rust, which would normally take care of this. I could make it non-generic instead.
